server/src/test/java/org/elasticsearch/common/util/CancellableSingleObjectCacheTests.java

```diff
@@ -0,0 +1,264 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.StepListener;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.tasks.TaskCancelledException;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class CancellableSingleObjectCacheTests extends ESTestCase {
+
+    public void testNoPendingRefreshIfAlreadyCancelled() {
+        final TestCache testCache = new TestCache();
+        final PlainActionFuture<Integer> future = new PlainActionFuture<>();
+        testCache.get("foo", () -> true, future);
+        testCache.assertPendingRefreshes(0);
+        assertTrue(future.isDone());
+        expectThrows(ExecutionException.class, TaskCancelledException.class, future::get);
+    }
+
+    public void testListenersCompletedByRefresh() {
+        final TestCache testCache = new TestCache();
+
+        // The first get() calls the refresh function
+        final PlainActionFuture<Integer> future0 = new PlainActionFuture<>();
+        testCache.get("foo", () -> false, future0);
+        testCache.assertPendingRefreshes(1);
+
+        // The second get() with a matching key does not refresh again
+        final PlainActionFuture<Integer> future1 = new PlainActionFuture<>();
+        testCache.get("foo", () -> false, future1);
+        assertFalse(future0.isDone());
+        assertFalse(future1.isDone());
+        testCache.assertPendingRefreshes(1);
+        testCache.completeNextRefresh("foo", 1);
+        assertThat(future0.actionGet(0L), equalTo(1));
+        assertThat(future0.actionGet(0L), sameInstance(future1.actionGet(0L)));
+
+        // A further get() call with a matching key re-uses the cached value
+        final PlainActionFuture<Integer> future2 = new PlainActionFuture<>();
+        final AtomicBoolean future2Cancelled = new AtomicBoolean();
+        testCache.get("foo", future2Cancelled::get, future2);
```

Comment:
I think `future2Cancelled::get` should just be `() -> false`?

Reply:
Well spotted, a leftover from an earlier iteration. Fixed in 927bcf5968b558e3d1a75ff85d788955539610a5.